### DISCUSSION TITLE
Easier way to apply commands to variables

### DISCUSSION THREAD

**Discussion by mikesimps** [igor-petruk/oatbar#18] [Ideas]
Maybe I'm missing something obvious, but it would be great if there was an way to do simple mathematical expressions on the output of a block. In my example, I just want to convert bits/sec into Mbps from the oatbar-stats variable. This doesn't work, but it would be cool if it did (or some other similar syntax).

<img width="367" alt="image" src="https://github.com/user-attachments/assets/cf68cc3d-9f3c-4aec-9773-0539de0c2e9c">

**Comment by igor-petruk** [comment]
Hi,

Thank you for the feedback!

My idea was to use number blocks for numbers https://oatbar.app/configuration/block.html#number-block with `number_type="bytes"`

In that case you could use just the variable inside of the value, but use `output_format` with `... {value}` to actually format the text. In that case units will be converted and adjusted automatically and you won't have to guess which order of magnitude your values have. In `output_format` you can use [`align` filter](https://oatbar.app/configuration/variable.html#filters) so the text does not jump around

 Please let me know if this works.

That being said, I do want to reach the least clunky way to configure these things, so if the result is not satisfactory with this approach, we can see what else is missing in number blocks to achieve ideal formatting.

Adding math expressions could be done, but still haven't decided how exactly. For example should integrate some library? If, yes, which one? Or should I build again something via Unix pipes so you can use your own choice of expression language. There are a few undecided questiosn there.

Thanks!

**Reply by mikesimps** [reply]
No, this is great for this use case, sorry I missed it! I will say I had a bit of trouble piecing everything together from the documentation, but it was technically there. I would recommend adding value and output_format to your example to make it more complete. This is my working config:

```toml
[[block]]
name="network"
type="number"
number_type="bytes"
min_value="0"
value="${stats:net.enp1s0.rx_per_sec}"
output_format="󰲝  ${value}"
```

Is there any way that I could "lock in" the byte unit? Like if I only wanted to see MB/s (so it may round to zero most of the time). Is that possible or would I have to create something custom?

Thanks for your help and great job on the project. Looking forward to watching it evolve.

**Comment by igor-petruk** [comment]
Yeah, adding an ability to lock a unit is a good idea for a feature, I will work on it and I'll try to make the docs more clear. Thanks 👍

**Reply by mikesimps** [reply]
Thanks! Really not a high priority, I'm just trying to keep the amount of flicker/movement down so keeping things in a standard spacing with standard units is helpful, but not critical. Definitely need a system tray more than anything else (which is on your list).

**Reply by igor-petruk** [reply]
Yeah, that makes sense. Also if widgets are not moving, then only those widgets are getting redrawn, not the entire bar. So it also saves CPU cycles. You can certainly reduce flicker with `align` filter

Tray protocols are a mess. A new protocol was never adopted, so much that Polybar decided not to implement it. The current protocol is very old, XEmbed based, very hard to navigate. It won't work on wayland. I've started this work, but it moves slowly. Meanwhile I can recommend something like https://kolbusa.github.io/stalonetray/, but yeah, it is not the same as embedded tray.

**Reply by igor-petruk** [reply]
The new release has an SNI based tray made in a very interesting way: https://oatbar.app/configuration/cookbook/tray.html
